Keyboard firmware has been written!

whampson opened the pull request:
I've completed the Arduino firmware that sends off MIDI commands based on keypresses. The software is untested as of now, but it compiles and loads successfully onto my Arduino Uno. The software is also pretty simple and, should any bugs arise, will be easy to debug.

whampson committed:
Add design overview

Commit:
Design overview: Added C5 to natural notes list and reassigned black keys to pins A1 - A5

Commit:
Fixed port numbers

Commit:
Moved MIDI library up one directory

Commit:
Defined MIDI note IDs and created Note struct

Commit:
Created Key class

Commit:
MIDI notes should now transmit when a key is pressed! (untested)

Commit:
Merge branch 'master' of https://github.com/SIGMusic/Grand-Piano into thehambone93/firmware

Conflicts:
	firmware/README.md

Commit:
Renamed 'm_midiID' to 'm_id'

Commit:
Removed ability to send MIDI command from within class

All MIDI operations are now handled by the main loop; added setPressed() function to record whether a key is being pressed; added getNote() function to retrieve the note that corresponds to the key

Commit:
Main loop now sends MIDI commands directly based on whether a key is being pressed

ajcord commented on firmware/midi_controller/midi_controller.ino:
```diff
@@ -1,30 +1,75 @@
-        MIDI.sendNoteOff(40, 0, 1);       // Stop the note!
+    for (int i = 0; i < NUM_KEYS; i++)
+    {
+        Key* pKey = &keys[i];
```
Is there a reason this is a pointer? Seems clearer to just access it directly.

whampson replied:
You're right, I'm not sure why I made it a pointer.

ajcord commented on firmware/midi_controller/midi_controller.ino:
```diff
@@ -1,30 +1,75 @@
+Key keyGSharp4(noteGSharp4, A4);
+Key keyASharp4(noteASharp4, A5);
+
+Key keys[] { keyC4, keyD4, keyE4, keyF4, keyG4, keyA4, keyB4, keyC5,
```
For safety, it would be a good idea to specify the size of `keys` as `NUM_KEYS`. That way, if you add/remove a key and forget to update `NUM_KEYS`, you get a compile error.

whampson replied:
Missed that, thanks!

ajcord commented on firmware/midi_controller/Note.hpp:
```diff
@@ -0,0 +1,31 @@
+#define MIDI_GSHARP4 56
+#define MIDI_ASHARP4 58
+
+struct Note
```
Personally, I would remove this data structure and split it up among other files. Velocity is a property of playback, not the note itself. Same with channel. They would really fit better being solely part of midi_controller.ino. ID makes more sense as a member of Key.

The MIDI_\* defines are good though, and it might be worth keeping this file around just for them. Alternatively, if you think they would fit better in midi_controller.ino, feel free to move them there.

whampson replied:
Yeah, I see what you mean. I'll give Key a `noteID` member and use that from the main loop. The MIDI_\* defines will probably remain in `Note.hpp` for now.

ajcord commented on firmware/midi_controller/midi_controller.ino:
```diff
@@ -1,30 +1,75 @@
+        int pinStatus = digitalRead(pKey->getPin());
+        if (pinStatus == 1 && !pKey->isPressed())
+        {
+            MIDI.sendNoteOn(note.m_id, note.m_velocity, note.m_channel);
```
Just hardcode `VELOCITY` and `MIDI_CHANNEL` here for now. If we ever implement velocity or multiple channels, we'll probably have to do it in this loop anyway.

whampson replied:
Good point!

ajcord commented on firmware/midi_controller/midi_controller.ino:
```diff
@@ -1,30 +1,75 @@
+        }
+        else if (pinStatus == 0 && pKey->isPressed())
+        {
+            MIDI.sendNoteOff(note.m_id, note.m_velocity, note.m_channel);
```
Same as above.

ajcord reviewed (changes requested):
Looks good overall! I left a few comments below.

Also, it might be a good idea to leave a small attribution notice in a comment at the top of each file. Nothing fancy, just 1 or 2 lines to say that it was made by SIGMusic and that its use is governed by the LICENSE file.

Commit:
Explicity defined size of 'keys' array

Commit:
loop(): changed the current key object from a pointer to a value

Commit:
Preparation for removal of Note struct

Using 'VELOCITY' and 'MIDI_CHANNEL' constants rather than the respective members of the Note struct in preparation for the removal of the Note struct

Commit:
Key class now contains a noteID member rather than a Note struct

'm_noteID' holds the MIDI ID of the note to be played when the key is pressed. Preparation for removal of Note struct

Commit:
Note IDs are now accessed through Key objects

Commit:
Note struct has been removed

Commit:
Renamed 'Note.hpp' to 'Notes.hpp'

Commit:
Removed commented-out code

Commit:
Documented functions and classes

Commit:
Added MIT license header and Copyright notice

whampson commented:
The changes have been made as requested.
